### PR TITLE
fix(berry arg parsing)

### PR DESCRIPTION
**Pull request by chris-olszewski**
This PR adds:
 - A new e2e test that verifies that package scripts aren't getting extra -- passed to them. e.g. NPM, Yarn, PNPM<7 all expect a -- to separate package manager args from script args, but other do not.
 - A fix for berry where turbo was passing -- as an arg separator. (See https://github.com/vercel/turborepo/issues/1477)
 - Add pnpm6 as a package manager that we run e2e tests against
 - Move pnpm6/pnpm7 behavior split to the standard package manager abstraction

I'm not happy with how I'm going about getting pnpm@6.22.2 for the e2e tests which involves installing it to a tempdir and then adding that to `$PATH` whenever we shell out, but it's what I got working that didn't mess with any global installs.

### CODE REVIEW

**Comment by jaredpalmer** [comment]
Failures are not due to this. Have some kind of config mishap with versioning

File: cli/scripts/monorepo.ts [hunk 1]
```diff
@@ -184,6 +184,7 @@ importers:
           test: `${turboPath} run test`,
           lint: `${turboPath} run lint`,
           special: "echo root task",
+          args: 'node -e "console.log(process.argv)" --',
```
Comment (nathanhammond) [inline cli/scripts/monorepo.ts:187]: Can you help me understand why we should include `--` here? I'm assuming that it's there for for a reason and I'd like to have it be a _bit_ more fleshed-out to help with detection of issues.

`node -e "console.log(process.argv)" --before-separator=thing -- --after-separator=thing`

We can also add additional scripts for "has separator" and "doesn't have separator."

File: cli/internal/packagemanager/packagemanager.go [hunk 2]
```diff
@@ -36,6 +36,10 @@ type PackageManager struct {
 	// The directory in which package assets are stored by the Package Manager.
 	PackageDir string
 
+	// The separator that the Package Manger uses to identify arguments that
+	// should be passed through to the underlying script.
+	ArgSeparator []string
```
Comment (nathanhammond) [inline cli/internal/packagemanager/packagemanager.go:41]: I know this maps to the previous, but can this be changed to not be an array of strings?

File: cli/scripts/e2e/e2e.ts [hunk 3]
```diff
@@ -487,6 +491,43 @@ function runSmokeTests<T>(
...
 
+  suite(
+    `${npmClient} passes through correct args ${
+      options.cwd ? " from " + options.cwd : ""
+    }`,
+    async () => {
+      const result = getCommandOutputAsArray(
+        repo.turbo(
+          "run",
+          ["args", "--filter=//", "--", "--script-arg=42"],
```
Comment (nathanhammond) [inline cli/scripts/e2e/e2e.ts:502]: Good call on choosing `filter` as a name.

Comment (nathanhammond) [inline cli/scripts/e2e/e2e.ts:502]: Can we do quite a few more variations on this with varying argument patterns? That'll likely make the rest of the stuff here need to become utility functions of some sort.

I'm thinking mostly edge cases:
- `-- -- --`
- ``
- `-f --f ---f ----f`
- `--same -- --same`
- `--zero -- --one -- --two -- --three`
- `-------------------------`

In terms of test understandability, I'd love to be able to spec the input like this and then the output expected adjacent to it. (e.g. Go table tests)

We might should also create multiple scripts in `package.json` that log out `argv` with and without their own separator.

**Review by nathanhammond — CHANGES_REQUESTED** [review]
Mostly: lots and lots more test scenarios!

File: cli/scripts/monorepo.ts (hunk 1 above)
Comment (chris-olszewski) [reply cli/scripts/monorepo.ts:187]: I had the trailing `--` so that we could detect when turbo was passing an unnecessary separator since Node swallows the first `--`:
```
olszewski@chriss-mbp cli % node -e "console.log(JSON.stringify(process.argv))" -- foo
["/Users/olszewski/.nvm/versions/node/v16.16.0/bin/node","foo"]
olszewski@chriss-mbp cli % node -e "console.log(JSON.stringify(process.argv))" -- -- foo
["/Users/olszewski/.nvm/versions/node/v16.16.0/bin/node","--","foo"]
```

Comment (nathanhammond) [reply cli/scripts/monorepo.ts:187]: Wow, it's turtles all the way down. Can you add this information in as a comment?

I still would like to have a `before` and `after` sentinel argument here, just to monitor for behavior changes.

Comment (chris-olszewski) [reply cli/scripts/monorepo.ts:187]: Can you elaborate on having a `before`/`after` sentinel arguments for this script? 

Unsure if it relates but, once node encounters an arg it doesn't recognize it treats that and every arg after it as a pass through:
```
olszewski@chriss-mbp cli % node -e "console.log(JSON.stringify(process.argv))" before -- after             
["/Users/olszewski/.nvm/versions/node/v16.16.0/bin/node","before","--","after"]
olszewski@chriss-mbp cli % node -e "console.log(JSON.stringify(process.argv))" --zero-fill-buffers before         
["/Users/olszewski/.nvm/versions/node/v16.16.0/bin/node","before"]
olszewski@chriss-mbp cli % node -e "console.log(JSON.stringify(process.argv))"  before --zero-fill-buffers
["/Users/olszewski/.nvm/versions/node/v16.16.0/bin/node","before","--zero-fill-buffers"]
```

File: cli/scripts/e2e/e2e.ts (hunk 3 above)
Comment (chris-olszewski) [reply cli/scripts/e2e/e2e.ts:502]: > We might should also create multiple scripts in package.json that log out argv with and without their own separator.

Could you elaborate on this?

File: cli/internal/packagemanager/packagemanager.go (hunk 2 above)
Comment (chris-olszewski) [reply cli/internal/packagemanager/packagemanager.go:41]: What's the most Go-like way to represent this? The comment on the pnpm 6/7 split:
> // We are allowed to use nil here because argSeparator already has a type, so it's a typed nil,
> // vs if we tried to call "args = append(args, nil...)", which would not work. This could
> // just as easily be []string{}, but the style guide says to prefer nil for empty slices.

Would it be best practice/common to use the empty string as the nil value and check it before appending to the arg slice?

Comment (nathanhammond) [reply cli/internal/packagemanager/packagemanager.go:41]: The existing solution is too clever by half, but I understand now why it exists.

- String types can't be `nil` (only empty string)
- `[]string{}`'s "zero value" is `nil`.

I'm ... comfortable leaving it the way it is. 🤷‍♂️ The only other reasonable options are:
- check for "" (not ideal since it's an in-band sentinel)
- far more-invasive (a whole new struct just for arg separator, and a method to get it, guards at the `append`, and then it's all just a mess)

So, yeah, let's add more to the comment and call it done.

File: cli/internal/packagemanager/pnpm.go [hunk 4]
```diff
@@ -21,6 +22,10 @@ var nodejsPnpm = PackageManager{
 	Specfile:   "package.json",
 	Lockfile:   "pnpm-lock.yaml",
 	PackageDir: "node_modules",
+	// pnpm v7+ changed their handling of '--'. We no longer need to pass it to pass args to
+	// the script being run, and in fact doing so will cause the '--' to be passed through verbatim,
+	// potentially breaking scripts that aren't expecting it.
+	ArgSeparator: nil,
```
Comment (nathanhammond) [inline cli/internal/packagemanager/pnpm.go:28]: Let's add information about this being a typed nil (`[]string{}`) here.

File: cli/internal/run/run.go [hunk 5]
```diff
@@ -929,7 +914,7 @@ func (e *execContext) exec(ctx gocontext.Context, pt *nodes.PackageTask, deps da
 	argsactual := append([]string{"run"}, pt.Task)
 	if len(passThroughArgs) > 0 {
 		// This will be either '--' or a typed nil
-		argsactual = append(argsactual, e.argSeparator...)
+		argsactual = append(argsactual, e.packageManager.ArgSeparator...)
```
Comment (nathanhammond) [inline cli/internal/run/run.go:917]: I love that the entire change is this single line.

File: cli/scripts/e2e/e2e.ts [hunk 6]
```diff
@@ -487,6 +491,65 @@ function runSmokeTests<T>(
...
+
+        assert.match(
+          node,
+          "node",
+          `Expected node binary path (${node}) to contain 'node'`
+        );
+        assert.equal(args, passedArgs);
+      };
+
+      const tests = [
```
Comment (nathanhammond) [inline cli/scripts/e2e/e2e.ts:523]: This is perfect, thank you. ❤️

File: cli/scripts/monorepo.ts [hunk 7]
```diff
@@ -53,6 +58,13 @@ export class Monorepo {
     if (!fs.existsSync(this.nodeModulesPath)) {
       fs.mkdirSync(this.nodeModulesPath, { recursive: true });
     }
+
+    if (this.npmClient === "pnpm6") {
+      fs.mkdirSync(path.join(this.binDir, "node_modules"), { recursive: true });
+      execa.sync("npm", ["install", "--no-save", "pnpm@6.22.2"], {
+        cwd: this.binDir,
+      });
+    }
```
Comment (nathanhammond) [inline cli/scripts/monorepo.ts:67]: You're right, corepack is a better story.

File: cli/scripts/e2e/e2e.ts (hunk 3 above)
Comment (nathanhammond) [reply cli/scripts/e2e/e2e.ts]: This is really an unnecessary, overly-defensive idea. Basically, we have a script called `argv` in the root `package.json` now that is just an `argv` echo service.

We could create variations on that script that had other `--` setups to detect if _node itself_ changes the way it works.

But that'll realistically break _way_ more than us and I can't see that being something we truly need to test.

File: cli/scripts/monorepo.ts (hunk 1 above)
Comment (nathanhammond) [reply cli/scripts/monorepo.ts]: Good callout that `node` itself blows up here. Thanks for that! This is fine as is.

**Review by nathanhammond — APPROVED** [review]
This is a fixship!
- The only open things remaining are something trivial. (e.g. all comments, naming, light refactoring, or just ... unimportant)
- So, once you're happy with it, ship it!
